pkg/backendaction/backendaction.go

```diff
@@ -124,38 +129,151 @@ func (c *Conn) Health(ctx context.Context) (json.RawMessage, error) {
 }
 
 func requestWithPassword(ctx context.Context, method string, addr string, pw []byte, body io.Reader) (json.RawMessage, error) {
-	req, err := http.NewRequestWithContext(ctx, method, addr, body)
-	if err != nil {
-		return nil, fmt.Errorf("creating request to backend action service: %w", err)
-	}
-	creds := shared.BasicAuth{
-		Password: pw,
-	}
-	req.Header.Set("Content-Type", "application/json")
-	req.Header.Set(shared.AuthHeader, creds.EncPassword())
+	const maxRetries = 5
+	const backoffDelay = 4 * time.Second
+	const requestTimeout = 5 * time.Second // Individual request timeout
 
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		return nil, fmt.Errorf("sending request to backend action service at %q: %w", addr, err)
+	var originalBody []byte
+	if body != nil {
+		var err error
+		originalBody, err = io.ReadAll(body)
+		if err != nil {
+			return nil, fmt.Errorf("reading request body: %w", err)
+		}
 	}
-	defer resp.Body.Close()
 
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		body, err := io.ReadAll(resp.Body)
+	var lastErr error
+	for attempt := 0; attempt < maxRetries; attempt++ {
+		// Check if parent context is already done before starting attempt
+		if ctx.Err() != nil {
+			log.Warn().Msg("parent context done before retry attempt")
+			return nil, ctx.Err()
+		}
+
+		var bodyReader io.Reader
+		if originalBody != nil {
+			bodyReader = bytes.NewReader(originalBody)
+		}
+
+		// Create a fresh context for this individual request
+		// This gives each request its own timeout window
+		reqCtx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+		defer cancel() // Always clean up the request context
+
+		req, err := http.NewRequestWithContext(reqCtx, method, addr, bodyReader)
 		if err != nil {
-			body = []byte("[can not read body]")
+			return nil, fmt.Errorf("creating request: %w", err)
 		}
-		return nil, fmt.Errorf("got response %q: %q", resp.Status, body)
+
+		creds := shared.BasicAuth{Password: pw}
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set(shared.AuthHeader, creds.EncPassword())
+
+		log.Warn().Int("attempt", attempt+1).Int("max_retries", maxRetries).Msg("attempting request")
+
+		resp, err := http.DefaultClient.Do(req)
```

Comment:
Move this and the handling of success to another func and only keep the error handling in case of `isNetworkError` in the loop.